File: routes/valuation.py
"""
Freedom Trader Equity Multiple valuation.

Inputs (per stock):
  - 5 years of ROE %                      (latest first: roe1 .. roe5)
  - 5 years of Dividend Payout Ratio       (decimal 0..1; latest first: payout1 .. payout5)
  - Required return %, Total Equity ($M), Shares Outstanding (M), Current Price

Per-stat (Average / Median / MOS=median*0.9) computation:
  - Normalised ROE  (%)
  - Payout          (decimal)  — corresponding stat across 5 years
  - Distributed (%) = ROE * Payout
  - Reinvested  (%) = ROE - Distributed = ROE * (1 - Payout)
  - Equity Multiplier  = (Reinvested %)^2 / 100
  - Equity Per Share   = Total Equity / Shares Outstanding
  - Valuation $        = Equity Per Share * Equity Multiplier
  - Discount %         = (Valuation - Price) / Valuation * 100

Assessment uses MOS valuation: price <= MOS => UNDERVALUED, <= median => FAIR_VALUE, else OVERVALUED.
"""

import statistics
from datetime import datetime, date
from flask import Blueprint, request, jsonify
from db import get_db


bp = Blueprint("valuation", __name__)


def _floats(vals):
    out = []
    for v in vals:
        if v is None or v == "":
            continue
        try:
            out.append(float(v))
        except (TypeError, ValueError):
            continue
    return out


def calculate(p):
    roes = _floats([p.get(f"roe{i}") for i in range(1, 6)])
    payouts = _floats([p.get(f"payout{i}") for i in range(1, 6)])
    if not roes:
        return None, "at least one ROE year required"

    total_eq = float(p.get("total_equity_m") or 0)
    shares = float(p.get("shares_outstanding_m") or 0)
    current = float(p.get("current_price") or 0)
    if shares <= 0:
        return None, "shares outstanding required"

    eps_book = total_eq / shares  # equity per share (book value)

    roe_avg = statistics.mean(roes)
    roe_med = statistics.median(roes)
    roe_mos = roe_med * 0.90

    if payouts:
        payout_avg = statistics.mean(payouts)
        payout_med = statistics.median(payouts)
    else:
        payout_avg = 0.0
        payout_med = 0.0
    payout_mos = payout_med * 0.90

    def distrib(roe, payout):
        return roe * payout

    def reinv(roe, payout):
        return roe * (1.0 - payout)

    def mult(reinvested):
        return (reinvested ** 2) / 100.0

    def disc(v):
        return ((v - current) / v * 100) if v else 0.0

    cols = {}
    for name, roe, payout in (
        ("avg",    roe_avg, payout_avg),
        ("median", roe_med, payout_med),
        ("mos",    roe_mos, payout_mos),
    ):
        d = distrib(roe, payout)
        r = reinv(roe, payout)
        m = mult(r)
        v = eps_book * m
        cols[name] = {
            "roe": roe, "payout": payout,
            "distributed": d, "reinvested": r,
            "multiplier": m, "valuation": v, "discount_pct": disc(v),
        }

    if cols["mos"]["valuation"] > 0 and current <= cols["mos"]["valuation"]:
        assessment = "UNDERVALUED"
    elif cols["median"]["valuation"] > 0 and current <= cols["median"]["valuation"]:
        assessment = "FAIR_VALUE"
    else:
        assessment = "OVERVALUED"

    return {
        "equity_per_share": round(eps_book, 4),
        "roe_avg": round(roe_avg, 4),
        "roe_median": round(roe_med, 4),
        "roe_mos": round(roe_mos, 4),
        "payout_avg": round(payout_avg, 6),
        "payout_median": round(payout_med, 6),
        "payout_mos": round(payout_mos, 6),
        "distributed_avg": round(cols["avg"]["distributed"], 4),
        "distributed_median": round(cols["median"]["distributed"], 4),
        "distributed_mos": round(cols["mos"]["distributed"], 4),
        "reinvested_avg": round(cols["avg"]["reinvested"], 4),
        "reinvested_median": round(cols["median"]["reinvested"], 4),
        "reinvested_mos": round(cols["mos"]["reinvested"], 4),
        "multiplier_avg": round(cols["avg"]["multiplier"], 4),
        "multiplier_median": round(cols["median"]["multiplier"], 4),
        "multiplier_mos": round(cols["mos"]["multiplier"], 4),
        "valuation_avg": round(cols["avg"]["valuation"], 4),
        "valuation_median": round(cols["median"]["valuation"], 4),
        "valuation_mos": round(cols["mos"]["valuation"], 4),
        "discount_avg_pct": round(cols["avg"]["discount_pct"], 2),
        "discount_median_pct": round(cols["median"]["discount_pct"], 2),
        "discount_mos_pct": round(cols["mos"]["discount_pct"], 2),
        "assessment": assessment,
    }, None


@bp.route("/preview", methods=["POST"])
def preview():
    p = request.get_json(force=True)
    result, err = calculate(p)
    if err:
        return jsonify({"success": False, "error": err}), 400
    return jsonify({"success": True, "data": result})


@bp.route("", methods=["POST"])
def save():
    p = request.get_json(force=True)
    result, err = calculate(p)
    if err:
        return jsonify({"success": False, "error": err}), 400

    ticker = (p.get("ticker") or "").strip().upper()
    if not ticker:
        return jsonify({"success": False, "error": "ticker required"}), 400

    now = datetime.now().isoformat(timespec="seconds")
    vdate = p.get("valuation_date") or date.today().isoformat()

    with get_db() as db:
        cur = db.execute("""
            INSERT INTO valuations (
                ticker, company_name, valuation_date, current_price,
                roe1, roe2, roe3, roe4, roe5,
                payout1, payout2, payout3, payout4, payout5,
                required_return, total_equity_m, shares_outstanding_m,
                equity_per_share,
                roe_avg, roe_median, roe_mos,
                payout_avg, payout_median, payout_mos,
                distributed_avg, distributed_median, distributed_mos,
                reinvested_avg, reinvested_median, reinvested_mos,
                multiplier_avg, multiplier_median, multiplier_mos,
                valuation_avg, valuation_median, valuation_mos,
                discount_avg_pct, discount_median_pct, discount_mos_pct,
                assessment, created_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,
                      ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,
                      ?, ?, ?, ?, ?)
        """, (
            ticker, p.get("company_name") or "", vdate, p.get("current_price"),
            p.get("roe1"), p.get("roe2"), p.get("roe3"), p.get("roe4"), p.get("roe5"),
            p.get("payout1"), p.get("payout2"), p.get("payout3"), p.get("payout4"), p.get("payout5"),
            p.get("required_return") or 10.0,
            p.get("total_equity_m"), p.get("shares_outstanding_m"),
            result["equity_per_share"],
            result["roe_avg"], result["roe_median"], result["roe_mos"],
            result["payout_avg"], result["payout_median"], result["payout_mos"],
            result["distributed_avg"], result["distributed_median"], result["distributed_mos"],
            result["reinvested_avg"], result["reinvested_median"], result["reinvested_mos"],
            result["multiplier_avg"], result["multiplier_median"], result["multiplier_mos"],
            result["valuation_avg"], result["valuation_median"], result["valuation_mos"],
            result["discount_avg_pct"], result["discount_median_pct"], result["discount_mos_pct"],
            result["assessment"], now,
        ))
        vid = cur.lastrowid

        rid = p.get("research_id")
        if rid:
            db.execute(
                "UPDATE researched_stocks SET latest_valuation_id = ?, updated_at = ? WHERE id = ?",
                (vid, now, rid),
            )

        row = db.execute("SELECT * FROM valuations WHERE id = ?", (vid,)).fetchone()
    return jsonify({"success": True, "data": dict(row)})


@bp.route("/<ticker>", methods=["GET"])
def latest_for_ticker(ticker):
    with get_db() as db:
        row = db.execute(
            "SELECT * FROM valuations WHERE ticker = ? ORDER BY created_at DESC LIMIT 1",
            (ticker.upper(),),
        ).fetchone()
    return jsonify({"success": True, "data": dict(row) if row else None})

File: static/js/views/Valuation.js
import { post, get, fmtMoney, fmtNum, fmtPct, isoToday, assessmentClass } from "../utils.js";

const ROE_KEYS = ["roe1", "roe2", "roe3", "roe4", "roe5"];
const PAYOUT_KEYS = ["payout1", "payout2", "payout3", "payout4", "payout5"];
const NUMERIC_KEYS = [
  "current_price",
  ...ROE_KEYS, ...PAYOUT_KEYS,
  "required_return", "total_equity_m", "shares_outstanding_m",
];

export const Valuation = {
  props: ["ticker"],
  data() {
    return {
      form: this._initForm(),
      result: null,
      saving: false,
      saveError: null,
      prefilling: false,
      prefillNote: null,
      prefillError: null,
    };
  },
  computed: {
    canCompute() {
      const f = this.form;
      const hasRoe = ROE_KEYS.some(k => f[k] !== null && f[k] !== "");
      return hasRoe && Number(f.shares_outstanding_m) > 0 && Number(f.total_equity_m) >= 0;
    },
    canSave() { return this.canCompute && !!this.form.ticker; },
  },
  watch: {
    form: { deep: true, handler() { this.recompute(); } },
  },
  methods: {
    _initForm() {
      const f = {
        ticker: "", company_name: "",
        valuation_date: isoToday(),
        current_price: null,
        required_return: 10,
        total_equity_m: null,
        shares_outstanding_m: null,
      };
      for (const k of ROE_KEYS) f[k] = null;
      for (const k of PAYOUT_KEYS) f[k] = null;
      return f;
    },
    payload() {
      const p = { ...this.form };
      NUMERIC_KEYS.forEach(k => {
        if (p[k] === "" || p[k] == null) p[k] = null;
        else p[k] = Number(p[k]);
      });
      // Convert payout from % input (e.g. 27) to decimal (0.27) for the API
      for (const k of PAYOUT_KEYS) {
        if (p[k] != null) p[k] = p[k] / 100;
      }
      p.ticker = (p.ticker || "").toUpperCase();
      const rid = this.$route?.query?.from_research;
      if (rid) p.research_id = Number(rid);
      return p;
    },
    backToResearch() {
      const rid = this.$route?.query?.from_research;
      if (rid) {
        this.$router.push(`/research/${rid}`);
      } else {
        this.$router.push("/research");
      }
    },
    async recompute() {
      if (!this.canCompute) { this.result = null; return; }
      try {
        this.result = await post("/api/valuation/preview", this.payload());
      } catch (e) {
        this.result = null;
      }
    },
    async saveAndReturn() {
      if (!this.canSave) return;
      this.saving = true;
      this.saveError = null;
      try {
        await post("/api/valuation", this.payload());
        this.backToResearch();
      } catch (e) {
        this.saveError = e.message;
      } finally {
        this.saving = false;
      }
    },
    async loadLatest() {
      if (!this.form.ticker) return;
      try {
        const data = await get(`/api/valuation/${this.form.ticker.toUpperCase()}`);
        if (data) {
          const next = this._initForm();
          next.ticker = data.ticker;
          next.company_name = data.company_name || "";
          next.valuation_date = data.valuation_date || isoToday();
          next.current_price = data.current_price;
          next.required_return = data.required_return ?? 10;
          next.total_equity_m = data.total_equity_m;
          next.shares_outstanding_m = data.shares_outstanding_m;
          for (const k of ROE_KEYS) next[k] = data[k] ?? null;
          // DB stores payout as decimal; convert to % for display
          for (const k of PAYOUT_KEYS) next[k] = data[k] != null ? data[k] * 100 : null;
          this.form = next;
        }
      } catch (e) { console.error(e); }
      await this.prefillCompany(false);
    },
    async prefillCompany(force) {
      const t = (this.form.ticker || "").trim().toUpperCase();
      if (!t) return;
      try {
        const data = await get(`/api/company/${t}${force ? "?refresh=1" : ""}`);
        if (!data) return;
        if (force || !this.form.company_name) this.form.company_name = data.company_name || this.form.company_name;
      } catch (e) { console.error(e); }
    },
    // Full SEC/EDGAR prefill of the valuation inputs (ROE, payout, equity, shares, price).
    async prefillFinancials() {
      const t = (this.form.ticker || "").trim().toUpperCase();
      if (!t) return;
      this.prefilling = true;
      this.prefillError = null;
      this.prefillNote = null;
      try {
        const d = await get(`/api/prefill/${t}`);
        if (d.company_name && !this.form.company_name) this.form.company_name = d.company_name;
        const v = d.valuation || {};
        const keys = [
          "current_price", "required_return", "total_equity_m", "shares_outstanding_m",
          ...ROE_KEYS, ...PAYOUT_KEYS,
        ];
        for (const k of keys) {
          if (k in v && v[k] != null) this.form[k] = v[k];
        }
        // Force the results to compute now — don't rely solely on the form
        // watcher firing after a batch of programmatic assignments.
        await this.recompute();
        if (d.financial) {
          this.prefillNote = `⚠ Financial sector (${d.sic_desc}) — equity-multiple doesn't apply; values are reference only.`;
        } else if (d.veto) {
          this.prefillNote = `⚠ ${d.veto_reason}`;
        } else {
          this.prefillNote = `Prefilled from SEC EDGAR · shares [${d.shares_src || "—"}], price [${d.price_src || "—"}]. Verify before saving.`;
        }
      } catch (e) {
        this.prefillError = e.message;
      } finally {
        this.prefilling = false;
      }
    },
    fmtRoe(v) { return v == null ? "—" : `${Number(v).toFixed(2)}%`; },
    // Clear the inputs (ROE / payout / equity / shares / price) if the prefilled
    // numbers look wrong. Keeps ticker + company so you can re-Prefill or re-enter.
    resetForm() {
      const t = this.form.ticker, name = this.form.company_name;
      this.form = this._initForm();
      this.form.ticker = t;
      this.form.company_name = name;
      this.result = null;
      this.prefillNote = null;
      this.prefillError = null;
    },
  },
  mounted() {
    if (this.ticker) {
      this.form.ticker = this.ticker.toUpperCase();
      this.loadLatest();
    }
  },
  template: `
    <div>
      <div class="toolbar">
        <h1 style="margin: 0;">Equity Multiple Valuation</h1>
        <div class="spacer"></div>
        <span v-if="saveError" class="text-red" style="font-size: 0.85rem; margin-right: 0.5rem;">{{ saveError }}</span>
        <button class="btn-primary" :disabled="!canSave || saving" @click="saveAndReturn">
          {{ saving ? 'Saving…' : 'Save & Return' }}
        </button>
        <button @click="resetForm()" style="margin-left: 0.5rem;" title="Clear ROE / payout / equity / shares / price">Reset</button>
        <button @click="backToResearch()" style="margin-left: 0.5rem;">Cancel</button>
      </div>

      <div class="card" v-if="prefillError" style="border-left: 4px solid var(--red); padding: 0.6rem 0.9rem;">
        <span class="text-red">Prefill: {{ prefillError }}</span>
      </div>
      <div class="card" v-if="prefillNote" style="border-left: 4px solid var(--accent); padding: 0.6rem 0.9rem; font-size: 0.85rem;">
        {{ prefillNote }}
      </div>

      <div class="grid-2">
        <div class="card">
          <h3>Inputs</h3>
          <div class="grid-2">
            <div class="field">
              <label>Stock Code</label>
              <input type="text" v-model="form.ticker" @blur="loadLatest()" placeholder="ADBE">
            </div>
            <div class="field">
              <label>Company <button type="button" class="btn-ghost" style="font-size: 0.75rem; padding: 0 0.4rem; margin-left: 0.4rem;" :disabled="!form.ticker || prefilling" @click="prefillFinancials()" title="Auto-fill ROE / payout / equity / shares / price from SEC EDGAR">{{ prefilling ? 'Prefilling…' : 'Prefill' }}</button></label>
              <input type="text" v-model="form.company_name" placeholder="Adobe Inc.">
            </div>
            <div class="field">
              <label>Date</label>
              <input type="date" v-model="form.valuation_date">
            </div>
            <div class="field">
              <label>Current Stock Price</label>
              <input type="number" step="0.01" v-model.number="form.current_price">
            </div>
          </div>

          <h3 style="margin-top: 1rem;">Return on Equity (5 years)</h3>
          <div class="grid-5">
            <div class="field">
              <label>Year 1 (latest)</label>
              <input type="number" step="0.01" v-model.number="form.roe1" placeholder="%">
            </div>
            <div class="field">
              <label>Year 2</label>
              <input type="number" step="0.01" v-model.number="form.roe2" placeholder="%">
            </div>
            <div class="field">
              <label>Year 3</label>
              <input type="number" step="0.01" v-model.number="form.roe3" placeholder="%">
            </div>
            <div class="field">
              <label>Year 4</label>
              <input type="number" step="0.01" v-model.number="form.roe4" placeholder="%">
            </div>
            <div class="field">
              <label>Year 5</label>
              <input type="number" step="0.01" v-model.number="form.roe5" placeholder="%">
            </div>
          </div>

          <h3 style="margin-top: 1rem;">Dividend Payout Ratio (5 years)</h3>
          <p class="text-muted" style="margin: -0.25rem 0 0.5rem;">As percentages: 0 = no dividend, 30 = 30% paid out.</p>
          <div class="grid-5">
            <div class="field">
              <label>Year 1</label>
              <input type="number" step="0.1" v-model.number="form.payout1" placeholder="%">
            </div>
            <div class="field">
              <label>Year 2</label>
              <input type="number" step="0.1" v-model.number="form.payout2" placeholder="%">
            </div>
            <div class="field">
              <label>Year 3</label>
              <input type="number" step="0.1" v-model.number="form.payout3" placeholder="%">
            </div>
            <div class="field">
              <label>Year 4</label>
              <input type="number" step="0.1" v-model.number="form.payout4" placeholder="%">
            </div>
            <div class="field">
              <label>Year 5</label>
              <input type="number" step="0.1" v-model.number="form.payout5" placeholder="%">
            </div>
          </div>

          <h3 style="margin-top: 1rem;">Equity</h3>
          <div class="grid-3">
            <div class="field">
              <label>Required Return %</label>
              <input type="number" step="0.01" v-model.number="form.required_return">
            </div>
            <div class="field">
              <label>Total Equity ($ M)</label>
              <input type="number" step="0.01" v-model.number="form.total_equity_m">
            </div>
            <div class="field">
              <label>Shares Outstanding (M)</label>
              <input type="number" step="0.01" v-model.number="form.shares_outstanding_m">
            </div>
          </div>
        </div>

        <div class="card">
          <h3>Results</h3>
          <div v-if="!result" class="empty">Enter at least one ROE year, total equity, and shares to compute.</div>
          <div v-else>
            <div class="grid-3">
              <div class="val-cell">
                <div class="label">Equity Per Share</div>
                <div class="num">{{ fmtMoney(result.equity_per_share) }}</div>
              </div>
              <div class="val-cell">
                <div class="label">Current Price</div>
                <div class="num">{{ fmtMoney(form.current_price) }}</div>
              </div>
              <div class="val-cell">
                <div class="label">Assessment</div>
                <div><span :class="assessmentClass(result.assessment)">{{ result.assessment }}</span></div>
              </div>
            </div>

            <div class="table-wrap" style="margin-top: 1rem;"><table class="table">
              <thead>
                <tr><th></th><th class="num">Average</th><th class="num">Median</th><th class="num">MOS (10% disc)</th></tr>
              </thead>
              <tbody>
                <tr>
                  <td>Normalised ROE</td>
                  <td class="num">{{ fmtRoe(result.roe_avg) }}</td>
                  <td class="num">{{ fmtRoe(result.roe_median) }}</td>
                  <td class="num">{{ fmtRoe(result.roe_mos) }}</td>
                </tr>
                <tr>
                  <td>Distributed</td>
                  <td class="num">{{ fmtRoe(result.distributed_avg) }}</td>
                  <td class="num">{{ fmtRoe(result.distributed_median) }}</td>
                  <td class="num">{{ fmtRoe(result.distributed_mos) }}</td>
                </tr>
                <tr>
                  <td>Reinvested</td>
                  <td class="num">{{ fmtRoe(result.reinvested_avg) }}</td>
                  <td class="num">{{ fmtRoe(result.reinvested_median) }}</td>
                  <td class="num">{{ fmtRoe(result.reinvested_mos) }}</td>
                </tr>
                <tr>
                  <td>Equity Multiplier</td>
                  <td class="num">{{ fmtNum(result.multiplier_avg) }}</td>
                  <td class="num">{{ fmtNum(result.multiplier_median) }}</td>
                  <td class="num">{{ fmtNum(result.multiplier_mos) }}</td>
                </tr>
                <tr>
                  <td>Valuation</td>
                  <td class="num">{{ fmtMoney(result.valuation_avg) }}</td>
                  <td class="num">{{ fmtMoney(result.valuation_median) }}</td>
                  <td class="num">{{ fmtMoney(result.valuation_mos) }}</td>
                </tr>
                <tr>
                  <td>Discount / Overvalued %</td>
                  <td class="num" :class="result.discount_avg_pct >= 0 ? 'text-green' : 'text-red'">
                    {{ fmtPct(result.discount_avg_pct) }}
                  </td>
                  <td class="num" :class="result.discount_median_pct >= 0 ? 'text-green' : 'text-red'">
                    {{ fmtPct(result.discount_median_pct) }}
                  </td>
                  <td class="num" :class="result.discount_mos_pct >= 0 ? 'text-green' : 'text-red'">
                    {{ fmtPct(result.discount_mos_pct) }}
                  </td>
                </tr>
              </tbody>
            </table></div>
          </div>
        </div>
      </div>
    </div>
  `,
  setup() { return { fmtMoney, fmtNum, fmtPct, assessmentClass }; },
};
